From 725136ad27663602075ab8e87ba308ef0b0d6676 Mon Sep 17 00:00:00 2001
From: Martin Storsjo <martin@martin.st>
Date: Mon, 22 Oct 2018 23:41:41 +0300
Subject: [PATCH 7/8] Expand response files

On Windows, meson always uses response files for passing arguments
(since there's a real risk of overflowing the max command line length).

Expand them to a normal command line within gas-preprocessor, where
we assume that the argument list won't be too long.
---
 gas-preprocessor.pl | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/gas-preprocessor.pl b/gas-preprocessor.pl
index 0e8529c..123bf32 100755
--- a/gas-preprocessor.pl
+++ b/gas-preprocessor.pl
@@ -56,6 +56,27 @@ sub usage() {
     print $usage_str;
 }
 
+my $index = 0;
+while ($index <= $#ARGV) {
+    if ($ARGV[$index] =~ /^@/) {
+        my $file = $ARGV[$index];
+        splice(@ARGV, $index, 1);
+        $file =~ s/^@//;
+        open(INPUT, "<", $file) || die "Error reading response file $file";
+        while (<INPUT>) {
+            my @args = split(" ", $_);
+            foreach my $arg (@args) {
+                $arg =~ s/^"//;
+                $arg =~ s/"$//;
+                splice(@ARGV, $index, 0, $arg);
+            }
+        }
+        close(INPUT);
+	next;
+    }
+    $index++;
+}
+
 while (@ARGV) {
     my $opt = shift;
 
@@ -109,7 +130,7 @@ if ($as_type eq "armasm") {
     $preprocess_c_cmd[0] = "cpp";
 
     # Remove -ignore XX parameter pairs from preprocess_c_cmd
-    my $index = 1;
+    $index = 1;
     while ($index < $#preprocess_c_cmd) {
         if ($preprocess_c_cmd[$index] eq "-ignore" and $index + 1 < $#preprocess_c_cmd) {
             splice(@preprocess_c_cmd, $index, 2);
@@ -160,7 +181,7 @@ if ((grep /^[-\/]c$/, @gcc_cmd) && !(grep /^-o/, @gcc_cmd) && !(grep /^[-\/Fo]/,
     }
 }
 # Remove the -o argument; if omitted, we by default preprocess to stdout.
-my $index = 1;
+$index = 1;
 while ($index < $#preprocess_c_cmd) {
     if ($preprocess_c_cmd[$index] eq "-o" or $preprocess_c_cmd[$index] =~ /^[-\/]Fo$/) {
         splice(@preprocess_c_cmd, $index, 2);
-- 
2.17.1

